### PR TITLE
CTIM Note entity Support

### PR DESCRIPTION
**Pull request by gbuisson**
> **Epic** https://github.com/advthreat/iroh/issues/7466
> Close  https://github.com/advthreat/iroh/issues/7467
Related: https://github.com/threatgrid/ctim/pull/395

Implement the Note Entity in CTIA:

- [x] CRUD routes
- [x] Note Store
- [x] CRUD TESTS
- [x] Bulk TESTS
- [x] Bundle TESTS
- [x] Migration TESTS

<a name="qa">[§](#qa)</a> QA
============================

The tests should be similar to the other CTIA CRUD APIs, and should respect the behavior as specified in the related design.
here are some samples queries

1. Create Notes

    
    ```http
    POST http://localhost:3000/ctia/note
    content-type: application/json
    
    {"entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
     "content": "This is a Note",
     "source": "Cisco Threat Response"
    }	      
    ```
    
    ```
    {
      "schema_version": "1.2.1",
      "content": "This is a Note",
      "type": "note",
      "source": "Cisco Threat Response",
      "id": "http://localhost:3000/ctia/note/note-96f6cdf8-c001-4fb8-a52d-8fc0b230c32d",
      "tlp": "green",
      "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
      "groups": [
        "Administrators"
      ],
      "timestamp": "2023-02-01T13:20:57.959Z",
      "owner": "Unknown"
    }
    ```
    
    ```http
    POST http://localhost:3000/ctia/note
    content-type: application/json
    
    {"entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
     "content": "This is another Note",
     "source": "Cisco Threat Response"
    }    	          
    ```

    ```
    {
      "schema_version": "1.2.1",
      "content": "This is another Note",
      "type": "note",
      "source": "Cisco Threat Response",
      "id": "http://localhost:3000/ctia/note/note-e2faad03-7b29-4c29-bfee-1fb5e04170ff",
      "tlp": "green",
      "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
      "groups": [
        "Administrators"
      ],
      "timestamp": "2023-02-01T13:20:59.432Z",
      "owner": "Unknown"
    }
    ```

2. PUT a Note

    Sample:  
    
    ```http
    PUT http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947
    content-type: application/json
    
    {"entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
     "content": "This is an updated Note",
     "source": "Cisco Threat Response"
    }
    ```
    
    ```
    {
      "schema_version": "1.2.1",
      "content": "This is an updated Note",
      "type": "note",
      "source": "Cisco Threat Response",
      "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
      "tlp": "green",
      "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
      "groups": [
        "Administrators"
      ],
      "timestamp": "2023-02-01T13:21:00.765Z",
      "owner": "Unknown"
    }
    ```

3.  PATCH (partial update) a Note

    ```http
    PATCH http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947
    content-type: application/json
    
    {
     "content": "This is a patched Note"
    }
    ```
    
    ```
    {
      "schema_version": "1.2.1",
      "content": "This is a patched Note",
      "type": "note",
      "source": "Cisco Threat Response",
      "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
      "tlp": "green",
      "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
      "groups": [
        "Administrators"
      ],
      "timestamp": "2023-02-01T13:21:00.765Z",
      "owner": "Unknown"
    }

4. GET a Note

    ```http
    GET http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947
    accept: application/json
    ```
    
    ```
    {
      "schema_version": "1.2.1",
      "content": "This is a patched Note",
      "type": "note",
      "source": "Cisco Threat Response",
      "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
      "tlp": "green",
      "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
      "groups": [
        "Administrators"
      ],
      "timestamp": "2023-02-01T13:21:00.765Z",
      "owner": "Unknown"
    }
    ```

5. Search Notes

    Sample searches:  
    
    ```http
    GET http://localhost:3000/ctia/note/search
    accept: application/json
    
    	
    ```
    
    ```
    [
      {
        "schema_version": "1.2.1",
        "content": "This is a Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-64b6b2a2-2f3e-496f-9989-1fb165975c70",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T09:45:59.748Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is another Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-ba46db2e-cdbc-48b2-853e-3a53ca71a7b0",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T09:47:58.627Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is a Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-96f6cdf8-c001-4fb8-a52d-8fc0b230c32d",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:20:57.959Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is another Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-e2faad03-7b29-4c29-bfee-1fb5e04170ff",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:20:59.432Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is a patched Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:21:00.765Z",
        "owner": "Unknown"
      }
    ]
    ```
    
    Search by Entity ID and order by timestamp desc  
    
    ```http
    GET http://localhost:3000/ctia/note/search?entity_id=http%3A%2F%2Flocalhost%3A3000%2Fctia%2Fincident%2Fincident-547dec6a-6bd3-4840-b3a1-8bb51726908c&sort_by=timestamp:desc
    accept: application/json
    
    	
    ```
    
    ```
    [
      {
        "schema_version": "1.2.1",
        "content": "This is a patched Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:21:00.765Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is another Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-e2faad03-7b29-4c29-bfee-1fb5e04170ff",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:20:59.432Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is a Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-96f6cdf8-c001-4fb8-a52d-8fc0b230c32d",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:20:57.959Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is another Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-ba46db2e-cdbc-48b2-853e-3a53ca71a7b0",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T09:47:58.627Z",
        "owner": "Unknown"
      },
      {
        "schema_version": "1.2.1",
        "content": "This is a Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-64b6b2a2-2f3e-496f-9989-1fb165975c70",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T09:45:59.748Z",
        "owner": "Unknown"
      }
    ]
    ```
    
    Content Fulltext search:  
    
    ```http
    GET http://localhost:3000/ctia/note/search?simple_query=patched
    accept: application/json
    ```
    
    ```
    [
      {
        "schema_version": "1.2.1",
        "content": "This is a patched Note",
        "type": "note",
        "source": "Cisco Threat Response",
        "id": "http://localhost:3000/ctia/note/note-c39fb90b-d234-41fd-8da9-97da67e12947",
        "tlp": "green",
        "entity_id": "http://localhost:3000/ctia/incident/incident-547dec6a-6bd3-4840-b3a1-8bb51726908c",
        "groups": [
          "Administrators"
        ],
        "timestamp": "2023-02-01T13:21:00.765Z",
        "owner": "Unknown"
      }
    ]
    ```

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: CTIA Note Entity Support
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: resources/ctia-default.properties [hunk 1]
```diff
@@ -68,9 +68,9 @@ ctia.store.es.default.port=9205
 #ctia.store.es.default.host=es-storage.int.iroh.site
 #ctia.store.es.default.port=443
 #ctia.store.es.default.protocol=https
-#ctia.store.es.default.auth.type=basic-auth
-#ctia.store.es.default.auth.params.user=elastic
-#ctia.store.es.default.auth.params.pwd=ductile
+ctia.store.es.default.auth.type=basic-auth
+ctia.store.es.default.auth.params.user=elastic
+ctia.store.es.default.auth.params.pwd=ductile
```
Comment (ereteog) [inline resources/ctia-default.properties:73]: it still a draft PR, but we must keep this commented, default properties are merged with provided ones.

File: src/ctia/entity/note.clj [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+(def note-mapping
+  {"note"
+   {:dynamic false
+    :properties
+    (merge
+     em/base-entity-mapping
+     em/sourcable-entity-mapping
+     em/stored-entity-mapping
+     {:entity_id em/searchable-token
```
Comment (ereteog) [inline src/ctia/entity/note.clj:22]: ids should only user `token` mapping (searchable adds an additional text field to have some fields indexed both as tokens and text, we introduced it for sources to add this additional field without reindexing everything)

File: src/ctia/entity/note.clj [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  [services :- APIHandlerServices]
+  (st/merge
+   routes.common/PagingParams
+   routes.common/BaseEntityFilterParams
+   routes.common/SourcableEntityFilterParams
+   routes.common/SearchableEntityParams
+   note-schemas/NoteFieldsParam
+   (st/optional-keys
+    {:entity_id s/Str
+     :content s/Str})))
```
Comment (ereteog) [inline src/ctia/entity/note.clj:71]: services is not used (and you can move it to the schemas ns)
```suggestion
(s/defschema NoteSearchParams
  (st/merge
   routes.common/PagingParams
   routes.common/BaseEntityFilterParams
   routes.common/SourcableEntityFilterParams
   routes.common/SearchableEntityParams
   note-schemas/NoteFieldsParam
   (st/optional-keys
    {:entity_id s/Str
     :content s/Str}))
```

File: test/ctia/task/migration/migrate_es_stores_test.clj [hunk 4]
```diff
@@ -142,6 +143,7 @@
     (random-updates app bulk-res-1 (/ updates-nb 2))
     (random-updates app bulk-res-2 (/ updates-nb 2))))
 
+
```
Comment (ereteog) [inline test/ctia/task/migration/migrate_es_stores_test.clj:146]:
```suggestion
```

**Review by ereteog — APPROVED** [review]
LGTM

File: src/ctia/entity/note.clj [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+(def note-mapping
+  {"note"
+   {:dynamic false
+    :properties
+    (merge
+     em/base-entity-mapping
+     em/sourcable-entity-mapping
+     em/stored-entity-mapping
+     {:entity_id em/token
+      :content em/searchable-token
```
Comment (ereteog) [inline src/ctia/entity/note.clj:23]: Content should be text only.
We will not sort on `content`.

Comment (gbuisson) [reply src/ctia/entity/note.clj]: done

File: src/ctia/entity/note.clj (hunk 2 above)
Comment (gbuisson) [reply src/ctia/entity/note.clj]: done

File: src/ctia/entity/note.clj (hunk 3 above)
Comment (gbuisson) [reply src/ctia/entity/note.clj]: I refactored the ns and moved everything schema related to the relevant ns

**Comment by trcollier** [comment]
QA verified in TEST release v1.114

**Comment by trcollier** [comment]
QA verified in release v1.114

**Pull request merged**